certbot/src/certbot/_internal/renewal.py

```diff
@@ -322,6 +327,68 @@ def should_renew(config: configuration.NamespaceConfig, lineage: storage.Renewab
     display_util.notify("Certificate not yet due for renewal")
     return False
 
+def should_autorenew(lineage: storage.RenewableCert, acme: acme_client.ClientV2) -> bool:
+    """Should we now try to autorenew the most recent cert version?
+
+    This is a policy question and does not only depend on whether
+    the cert is expired. (This considers whether autorenewal is
+    enabled, whether the cert is revoked, and whether the time
+    interval for autorenewal has been reached.)
+
+    Note that this examines the numerically most recent cert version,
+    not the currently deployed version.
+
+    :returns: whether an attempt should now be made to autorenew the
+        most current cert version in this lineage
+    :rtype: bool
+
+    """
+    if lineage.autorenewal_is_enabled():
+        cert = lineage.version("cert", lineage.latest_common_version())
+
+        # Consider whether to attempt to autorenew this cert now
+        renewal_time = None
+        with open(cert, 'rb') as f:
+            cert_pem = f.read()
+        renewal_time, _ = acme.renewal_time(cert_pem)
+
+        now = datetime.datetime.now(datetime.timezone.utc)
+
+        if renewal_time and now > renewal_time:
+            return True
+
+        # Renewals on the basis of revocation
+        if lineage.ocsp_revoked(lineage.latest_common_version()):
+            logger.debug("Should renew, certificate is revoked.")
+            return True
+
+        notBefore = crypto_util.notBefore(cert)
+        notAfter = crypto_util.notAfter(cert)
+        lifetime = notAfter - notBefore
+
+        config_interval = lineage.configuration.get("renew_before_expiry")
+        if (config_interval is not None and
+            notAfter < storage.add_time_interval(now, config_interval)):
+            logger.debug("Should renew, less than %s before certificate "
+                            "expiry %s.", config_interval,
+                            notAfter.strftime("%Y-%m-%d %H:%M:%S %Z"))
+            return True
+
+        # No config for "renew_before_expiry", provide default behavior.
+        # For most certs, renew with 1/3 of certificate lifetime remaining.
+        # For short lived certificates, renew at 1/2 of certificate lifetime.
+        default_interval = lifetime / 3
+        if lifetime.total_seconds() < 10 * 86400:
+            default_interval = lifetime / 2
+        remaining_time = notAfter - now
+        if remaining_time < default_interval:
+            logger.debug("Should renew, less than %ss before certificate "
+                            "expiry %s.", default_interval,
+                            notAfter.strftime("%Y-%m-%d %H:%M:%S %Z"))
+            return True
```

Comment:
isn't this obviated by the logic in `acme.renewal_time()`?

Reply:
Good point, I removed the logic here that duplicates what's in `renewal_time()`. I also updated the comment on `should_autorenew()` to describe the functionality a little more clearly, and tweaked `test_time_interval_judgements` to reflect the new split of responsibilities.